pam/internal/adapter/nativemodel.go

```diff
@@ -0,0 +1,633 @@
+			{id: "continue", label: fmt.Sprintf("Proceed with %s", authMode)},
+		}
+		if buttonLabel := m.uiLayout.GetButton(); buttonLabel != "" {
+			choices = append(choices, choicePair{id: "button", label: buttonLabel})
+		}
+
+		id, err := m.promptForChoice(authMode, choices, "Select action")
+		if errors.Is(err, errGoBack) {
+			return m.goBackCommand()
+		}
+		if errors.Is(err, errEmptyResponse) {
+			return sendEvent(nativeChallengeRequested{})
+		}
+		if err != nil {
+			return sendPamError(err)
+		}
+		if id == "button" {
+			return sendEvent(reselectAuthMode{})
+		}
+	}
+
+	var prompt string
+	if m.uiLayout.Label != nil {
+		prompt, _ = strings.CutSuffix(*m.uiLayout.Label, ":")
+	}
+
+	if prompt == "" {
+		switch m.uiLayout.GetEntry() {
+		case "digits":
+			fallthrough
+		case "digits_password":
+			prompt = "PIN"
+		case "chars":
+			prompt = "Value"
+		case "chars_password":
+			prompt = "Password"
+		}
+	}
+
+	instructions := "Leave the input field empty or insert '%[1]s' to cancel the request and go back"
+	if hasWait {
+		// Duplicating some contents here, as it will be better for translators once we've them
+		instructions = "Leave the input field empty to wait for other authentication method" +
+			"or insert '%[1]s' to go back"
+		if m.uiLayout.GetEntry() == "" {
+			instructions = "Leave the input field empty to wait for the authentication method" +
+				"or insert '%[1]s' to go back"
+		}
+	}
+
+	err := m.sendInfo(instructions, nativeCancelKey)
+	if err != nil {
+		return sendPamError(err)
+	}
+
+	challenge, err := m.promptForChallenge(prompt)
+	if errors.Is(err, errGoBack) {
+		return m.goBackCommand()
+	}
+	if errors.Is(err, errEmptyResponse) {
+		if hasWait {
+			return sendAuthWaitCommand()
+		}
+		err = nil
+	}
+	if err != nil {
+		return sendPamError(err)
+	}
+
+	return sendEvent(isAuthenticatedRequested{
+		item: &authd.IARequest_AuthenticationData_Challenge{Challenge: challenge},
+	})
+}
+
+// TestNativeAuthenticate/Authenticate_user_switching_to_local_broker
+
+func (m nativeModel) promptForChallenge(prompt string) (string, error) {
+	switch m.uiLayout.GetEntry() {
+	case "chars", "":
+		return m.promptForInput(pam.PromptEchoOn, prompt)
+	case "chars_password":
+		return m.promptForInput(pam.PromptEchoOff, prompt)
+	case "digits":
+		return m.promptForNumericInputAsString(pam.PromptEchoOn, prompt)
+	case "digits_password":
+		return m.promptForNumericInputAsString(pam.PromptEchoOff, prompt)
+	default:
+		return "", fmt.Errorf("Unhandled entry %q", m.uiLayout.GetEntry())
+	}
+}
+
+func (m nativeModel) handleQrCode() tea.Cmd {
+	qrCode, err := qrcode.New(m.uiLayout.GetContent(), qrcode.Medium)
+	if err != nil {
+		return sendEvent(pamError{
+			status: pam.ErrSystem,
+			msg:    fmt.Sprintf("Can't generate qrcode: %v", err),
+		})
+	}
+
+	err = m.sendInfo(m.uiLayout.GetLabel())
+	if err != nil {
+		return sendPamError(err)
+	}
+
+	err = m.sendInfo(qrCode.ToSmallString(false))
+	if err != nil {
+		return sendPamError(err)
+	}
```

Comment:
You're not using the error value later, so I think it's best to just switch to the shortened-if format